dp/generate_prsn.py

```diff
@@ -0,0 +1,272 @@
+    freezing = 0.0
+    if unit.lower() == 'k':
+        freezing = 273.15
+    return freezing
+
+
+def create_prsn_netcdf_from_source(src, dst):
+    '''Using a precipiation netCDF as a template copy over the data applicable
+       to prsn
+    '''
+    # Create the dimensions of the file
+    for name, dim in src.dimensions.items():
+        dst.createDimension(name, len(dim) if not dim.isunlimited() else None)
+
+    # Copy the globals
+    dst.setncatts({attr:src.getncattr(attr) for attr in src.ncattrs()})
+
+    # Create the variables in the file
+    for name, var in src.variables.items():
+        dst.createVariable(name, var.dtype, var.dimensions)
+
+        # Copy the variable attributes
+        dst.variables[name].setncatts({attr:var.getncattr(attr) for attr in var.ncattrs()})
+
+        # we will be replacing pr data with prsn data
+        if name != 'pr':
+            logger.debug('Copying {}'.format(name))
+            dst.variables[name][:] = src.variables[name][:]
+        else:
+            continue
+
+
+    # change pr metadata to prsn equivalents
+    dst.renameVariable('pr', 'prsn')
+    prsn_var = dst.variables['prsn']
+
+    prsn_var.standard_name = 'snowfall_flux'
+    prsn_var.long_name = 'Precipitation as Snow'
+
+    to_delete = ['original_name', 'comment']
+    for item in to_delete:
+        if hasattr(prsn_var, item):
+            prsn_var.delncattr(item)
+
+
+def create_filepath_from_source(source, new_var, outdir):
+    '''Using the source cmor_filename build a new output filepath with new_var
+       and output directory.
+    '''
+    variable, *rest = source.cmor_filename.split('_')
+
+    suffix = ''
+    for var in rest:
+        suffix += '_' + var
+
+    if not os.path.exists(os.path.dirname(outdir)):
+        os.makedirs(os.path.dirname(outdir))
+
+    return os.path.join(outdir, new_var + suffix)
+
+
+def has_required_vars(datasets, required_vars):
+    ''' Given a list of datasets and a list of required variables, ensure the
+        datasets contain all required variables.
+    '''
+    unique_vars = set()
+    for dataset in datasets:
+        for var in dataset.variables:
+            unique_vars.add(var)
+
+    for required_var in required_vars:
+        if required_var not in unique_vars:
+            logger.warning('Files do not contain required variables. required: {} have: {}'
+                           .format(required_vars, unique_vars))
+            return False
+
+    return True
+
+
+def matching_datasets(datasets):
+    '''Given a list of datasets, match important metadata vars to ensure the
+       datasets are compatible.
+    '''
+    required = {
+        'project': [],
+        'model': [],
+        'institute': [],
+        'experiment': [],
+        'ensemble_member': []
+    }
+
+    for dataset in datasets:
+        for attr in required.keys():
+            required[attr].append(getattr(dataset.metadata, attr))
```

Comment:
Nope, I think James was right. Comprehensions are to be preferred over explicit loops. Easier to read and understand, and less room for error:
```
required_attrs = {'project', 'model', 'institute', 'experiment', 'ensemble_member'}
required = { 
   attr: [getattr(dataset.metadata, attr) for dataset in datasets] 
      for attr in required_attrs
} 
```
You might want to eliminate the intermediate variable `required_attrs`, though it does provide some explanation by its name.